AWS Dynamo DB Item Unexpected EOF error

_This issue was originally opened by @dineshvm87 as hashicorp/terraform#19043. It was migrated here as a result of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Terraform issue tracker reserved for bug reports and feature requests. For general usage questions, please see: https://www.terraform.io/community.html.

If your issue relates to a specific Terraform provider, please open it in the provider's own repository. The index of providers is at https://github.com/terraform-providers .
-->

### Terraform Version
<!---
Run `terraform -v` to show the version, and paste the result between the ``` marks below.

If you are not running the latest version of Terraform, please try upgrading because your issue may have already been fixed.
-->

```
Terraform v0.11.8
+ provider.aws v1.39.0
```

### Terraform Configuration Files
<!--
Paste the relevant parts of your Terraform configuration between the ``` marks below.

For large Terraform configs, please use a service like Dropbox and share a link to the ZIP file. For security, you can also encrypt the files using our GPG public key.
-->

```hcl
variable "region" {
}
variable "aws_access_key" {
    type = "string"
}
variable "aws_secret_key" {
    type = "string"
}

provider "aws" {
  access_key = "${var.aws_access_key}"
  secret_key = "${var.aws_secret_key}"
  region     = "${var.region}"
}

resource "aws_dynamodb_table_item" "transit_connectivity" {
 table_name = "table_name"
 hash_key   = "aws:ca-central-1"
 item = <<ITEM
{
 "hub_id": {"S": "aws:ca-central-1"},
 "name": {"S": "name"}
}
ITEM
}
```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->

### Crash Output
<!--
If the console output indicates that Terraform crashed, please share a link to a GitHub Gist containing the output of the `crash.log` file.
-->
[https://github.com/dineshvm87/dynamodb-crash.log](url)
### Expected Behavior
<!--
Item details updated in the dynamoDB
-->
Item details updated in the dynamoDB
### Actual Behavior
<!--
Item details updated in the dynamoDB and output says terraform has crashed.
-->
Item details updated in the dynamoDB and output says terraform has crashed.
### Steps to Reproduce
<!--
-->
1. `terraform apply`
### Additional Context
<!--
There are no additional information. I am trying to update an item in the existing dynamo DB file.
-->

I am trying to update an item in the existing dynamo DB file. Item detail is in the ca-central-1 region but the dynamoDB table is in the us-west-2 region. Not sure if this is crashing due to two different regions. 
### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:

- hashicorp/terraform#6017

-->

I'm able to produce a crash by using an incorrect `hash_key`, e.g.

```hcl
terraform {
  required_version = "0.11.8"
}

provider "aws" {
  region  = "us-east-2"
  version = "1.39.0"
}

resource "aws_dynamodb_table" "test" {
  name           = "issue-6106-testing"
  hash_key       = "hub_id"
  read_capacity  = 1
  write_capacity = 1

  attribute {
    name = "hub_id"
    type = "S"
  }
}

resource "aws_dynamodb_table_item" "test" {
  hash_key   = "aws:ca-central-1"
  table_name = "${aws_dynamodb_table.test.name}"

  item = <<ITEM
{
 "hub_id": {"S": "aws:ca-central-1"},
 "name": {"S": "name"}
}
ITEM
}
```

```
panic: runtime error: invalid memory address or nil pointer dereference
2018-10-10T14:02:20.822-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: [signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x2bdaf63]
2018-10-10T14:02:20.822-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4:
2018-10-10T14:02:20.822-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: goroutine 191 [running]:
2018-10-10T14:02:20.822-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: github.com/terraform-providers/terraform-provider-aws/aws.buildDynamoDbTableItemId(0xc00048a220, 0x12, 0xc000180290, 0x10, 0x0, 0x0, 0xc0006e4ba0, 0x7c006c0, 0x0)
2018-10-10T14:02:20.822-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_dynamodb_table_item.go:248 +0x73
2018-10-10T14:02:20.822-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: github.com/terraform-providers/terraform-provider-aws/aws.resourceAwsDynamoDbTableItemCreate(0xc000b98a10, 0x37125e0, 0xc00040d500, 0xc000b98a10, 0x0)
2018-10-10T14:02:20.822-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/aws/resource_aws_dynamodb_table_item.go:82 +0x3ec
2018-10-10T14:02:20.823-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.(*Resource).Apply(0xc000a66e70, 0xc00096e370, 0xc0008aa300, 0x37125e0, 0xc00040d500, 0x100b601, 0xc00009cb80, 0x10bc6ec)
2018-10-10T14:02:20.823-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/resource.go:225 +0x351
2018-10-10T14:02:20.823-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema.(*Provider).Apply(0xc000b99490, 0xc00096e320, 0xc00096e370, 0xc0008aa300, 0xc000064c00, 0xc00009cbf0, 0x7c006c0)
2018-10-10T14:02:20.823-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/helper/schema/provider.go:283 +0x9c
2018-10-10T14:02:20.823-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/plugin.(*ResourceProviderServer).Apply(0xc0002c35e0, 0xc0008aa2a0, 0xc0003d2da0, 0x0, 0x0)
2018-10-10T14:02:20.823-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-aws/vendor/github.com/hashicorp/terraform/plugin/resource_provider.go:527 +0x57
2018-10-10T14:02:20.823-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: reflect.Value.call(0xc0000ace40, 0xc00072c030, 0x13, 0x3c5c16e, 0x4, 0xc00009cf18, 0x3, 0x3, 0xc00037c2c0, 0x0, ...)
2018-10-10T14:02:20.823-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: 	/opt/goenv/versions/1.11.1/src/reflect/value.go:447 +0x449
2018-10-10T14:02:20.823-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: reflect.Value.Call(0xc0000ace40, 0xc00072c030, 0x13, 0xc00061b718, 0x3, 0x3, 0x0, 0x0, 0x0)
2018-10-10T14:02:20.823-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: 	/opt/goenv/versions/1.11.1/src/reflect/value.go:308 +0xa4
2018-10-10T14:02:20.823-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: net/rpc.(*service).call(0xc0003cc5c0, 0xc000533770, 0xc000304b50, 0xc000304b70, 0xc000506300, 0xc0003f7be0, 0x321b340, 0xc0008aa2a0, 0x16, 0x321b380, ...)
2018-10-10T14:02:20.823-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: 	/opt/goenv/versions/1.11.1/src/net/rpc/server.go:384 +0x14e
2018-10-10T14:02:20.823-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: created by net/rpc.(*Server).ServeCodec
2018-10-10T14:02:20.823-0400 [DEBUG] plugin.terraform-provider-aws_v1.39.0_x4: 	/opt/goenv/versions/1.11.1/src/net/rpc/server.go:481 +0x47e
```

We'll want to know if this matches what you're seeing @dineshvm87. In the meantime, I think you may just need to update your `hash_key = "hub_id"` or adjust as appropriate. 👍

--- a/aws/resource_aws_dynamodb_table_item.go
+++ b/aws/resource_aws_dynamodb_table_item.go
@@ -240,46 +240,18 @@ func buildDynamoDbProjectionExpression(attrs map[string]*dynamodb.AttributeValue
 }
 
 func buildDynamoDbTableItemId(tableName string, hashKey string, rangeKey string, attrs map[string]*dynamodb.AttributeValue) string {
-	hashVal := attrs[hashKey]
+	id := []string{tableName, hashKey}
 
-	id := []string{
-		tableName,
-		hashKey,
-		base64Encode(hashVal.B),
+	if hashVal, ok := attrs[hashKey]; ok {
+		id = append(id, base64Encode(hashVal.B))
+		id = append(id, aws.StringValue(hashVal.S))
+		id = append(id, aws.StringValue(hashVal.N))
 	}
-
-	if hashVal.S != nil {
-		id = append(id, *hashVal.S)
-	} else {
-		id = append(id, "")
+	if rangeVal, ok := attrs[rangeKey]; ok && rangeKey != "" {
+		id = append(id, rangeKey, base64Encode(rangeVal.B))
+		id = append(id, aws.StringValue(rangeVal.S))
+		id = append(id, aws.StringValue(rangeVal.N))
 	}
-	if hashVal.N != nil {
-		id = append(id, *hashVal.N)
-	} else {
-		id = append(id, "")
-	}
-	if rangeKey != "" {
-		rangeVal := attrs[rangeKey]
-
-		id = append(id,
-			rangeKey,
-			base64Encode(rangeVal.B),
-		)
-
-		if rangeVal.S != nil {
-			id = append(id, *rangeVal.S)
-		} else {
-			id = append(id, "")
-		}
-
-		if rangeVal.N != nil {
-			id = append(id, *rangeVal.N)
-		} else {
-			id = append(id, "")
-		}
-
-	}
-
 	return strings.Join(id, "|")
 }
 

